Windows Warnings

```
src/XTModule.h:105:18: warning: format '%s' expects argument of type 'char*', but argument 6 has type 'const value_type*' {aka 'const wchar_t*'} [-Wformat=]
[834](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:834)
  105 |             INFO("[SurgeXTRack] storage::userDataPath = '%s'", storage->userDataPath.c_str());
[835](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:835)
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[836](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:836)
      |                                                                                           |
[837](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:837)
      |                                                                                           const value_type* {aka const wchar_t*}
[838](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:838)
/home/build/rack-plugin-toolchain/Rack-SDK-win/include/logger.hpp:14:105: note: in definition of macro 'INFO'
[839](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:839)
   14 | #define INFO(format, ...) rack::logger::log(rack::logger::INFO_LEVEL, __FILE__, __LINE__, __FUNCTION__, format, ##__VA_ARGS__)
[840](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:840)
      |                                                                                                         ^~~~~~
[841](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:841)
In file included from src/DelayLineByFreq.h:20,
[842](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:842)
                 from src/DelayLineByFreq.cpp:16:
[843](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:843)
src/XTModule.h:105:59: note: format string is defined here
[844](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:844)
  105 |             INFO("[SurgeXTRack] storage::userDataPath = '%s'", storage->userDataPath.c_str());
[845](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:845)
      |                                                          ~^
[846](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:846)
      |                                                           |
[847](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:847)
      |                                                           char*
[848](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:848)
      |
```

```cpp
src/VCO.h:650:10:   required from here
[1553](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1553)
src/VCO.h:707:14: warning: variable 'getBool' set but not used [-Wunused-but-set-variable]
[1554](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1554)
  707 |         auto getBool = [](auto *n, auto *str, auto def) {
[1555](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1555)
      |              ^~~~~~~
[1556](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1556)
src/VCO.h:713:14: warning: variable 'getInt' set but not used [-Wunused-but-set-variable]
[1557](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1557)
  713 |         auto getInt = [](auto *n, auto *str, auto def) -> int {
[1558](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1558)
      |              ^~~~~~
[1559](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1559)
src/VCO.h: In member function 'void sst::surgext_rack::vco::VCO<oscType>::readModuleSpecificJson(json_t*) [with int oscType = 7]':
[1560](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1560)
src/VCO.h:698:24: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 256 equals destination size [-Wstringop-truncation]
[1561](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1561)
  698 |                 strncpy(oscstorage->wavetable_display_name, dname.c_str(), 256);
[1562](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1562)
      |                 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[1563](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1563)
src/VCO.h:701:24: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 256 equals destination size [-Wstringop-truncation]
[1564](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1564)
  701 |                 strncpy(oscstorage_display->wavetable_display_name, dname.c_str(), 256);
[1565](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1565)
      |                 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[1566](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1566)
src/VCO.h: In member function 'void sst::surgext_rack::vco::VCO<oscType>::readModuleSpecificJson(json_t*) [with int oscType = 2]':
[1567](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1567)
src/VCO.h:698:24: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 256 equals destination size [-Wstringop-truncation]
[1568](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1568)
  698 |                 strncpy(oscstorage->wavetable_display_name, dname.c_str(), 256);
[1569](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1569)
      |                 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[1570](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1570)
src/VCO.h:701:24: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 256 equals destination size [-Wstringop-truncation]
[1571](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1571)
  701 |                 strncpy(oscstorage_display->wavetable_display_name, dname.c_str(), 256);
[1572](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1572)
      |                 ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[1573](https://github.com/surge-synthesizer/surge-rack/actions/runs/3282339625/jobs/5405611200#step:5:1573)
```

--- a/src/LFO.cpp
+++ b/src/LFO.cpp
@@ -33,20 +33,7 @@ struct LFOWidget : widgets::XTModuleWidget
 
 struct LFOTypeWidget : rack::app::ParamWidget, style::StyleParticipant
 {
-
     widgets::BufferedDrawFunctionWidget *bdw{nullptr};
-
-    /*
-     * auto logoSvg = SurgeInternal::getSurgeLogo();
-
-auto hn = logoSvg->handle;
-auto pt = panelTitle();
-for( auto s = hn->shapes; s; s = s->next )
-{
-    s->fill.color = (255 << 24) + (( (int)( pt.b * 255 ) ) << 16 )+ (( (int)( pt.g * 255 ) ) << 8) +
-(int)( pt.r * 255 );
-}
-     */
 
     static constexpr int nTypes{8};
     std::array<std::shared_ptr<rack::Svg>, nTypes> svgs;
@@ -149,9 +136,9 @@ for( auto s = hn->shapes; s; s = s->next )
 
         if (e.action == GLFW_PRESS)
         {
-            auto dSpan = box.size.x / (nTypes + 1);
+            auto dSpan = box.size.x / (nTypes);
             float which = std::floor(e.pos.x / dSpan);
-            module->paramQuantities[LFO::SHAPE]->setValue(which / n_lfo_types);
+            module->paramQuantities[LFO::SHAPE]->setValue(which / (n_lfo_types - 1));
             bdw->dirty = true;
             e.consume(this);
         }
@@ -172,6 +159,524 @@ for( auto s = hn->shapes; s; s = s->next )
         reloadGlyphs();
     }
     void onChange(const ChangeEvent &e) override { bdw->dirty = true; }
+};
+
+struct LFOWaveform : rack::Widget, style::StyleParticipant
+{
+    LFO *module{nullptr};
+    widgets::BufferedDrawFunctionWidget *bdwBG{nullptr}, *bdwLight{nullptr};
+    void setup()
+    {
+        bdwBG = new widgets::BufferedDrawFunctionWidget(rack::Vec(0, 0), box.size,
+                                                        [this](auto vg) { this->drawBG(vg); });
+        addChild(bdwBG);
+        bdwLight = new widgets::BufferedDrawFunctionWidgetOnLayer(
+            rack::Vec(0, 0), box.size, [this](auto vg) { this->drawLight(vg); });
+        addChild(bdwLight);
+        memset(tp, 0, n_scene_params * sizeof(pdata));
+    }
+
+    void onStyleChanged() override
+    {
+        bdwLight->dirty = true;
+        bdwBG->dirty = true;
+    }
+
+    void drawBG(NVGcontext *vg) {}
+
+    void drawLight(NVGcontext *vg)
+    {
+        if (!module)
+            return;
+        drawWaveform(vg);
+    }
+
+    pdata tp[n_scene_params];
+    void step() override
+    {
+        if (!module)
+            return;
+
+        auto lfodata = module->lfostorageDisplay;
+        auto storage = module->storage.get();
+
+        bool changed = false;
+        changed = changed || (tp[lfodata->delay.param_id_in_scene].i != lfodata->delay.val.i);
+        changed = changed || (tp[lfodata->attack.param_id_in_scene].i != lfodata->attack.val.i);
+        changed = changed || (tp[lfodata->hold.param_id_in_scene].i != lfodata->hold.val.i);
+        changed = changed || (tp[lfodata->decay.param_id_in_scene].i != lfodata->decay.val.i);
+        changed = changed || (tp[lfodata->sustain.param_id_in_scene].i != lfodata->sustain.val.i);
+        changed = changed || (tp[lfodata->release.param_id_in_scene].i != lfodata->release.val.i);
+
+        changed =
+            changed || (tp[lfodata->magnitude.param_id_in_scene].i != lfodata->magnitude.val.i);
+        changed = changed || (tp[lfodata->rate.param_id_in_scene].i != lfodata->rate.val.i);
+        changed = changed || (tp[lfodata->shape.param_id_in_scene].i != lfodata->shape.val.i);
+        changed =
+            changed || (tp[lfodata->start_phase.param_id_in_scene].i != lfodata->start_phase.val.i);
+        changed = changed || (tp[lfodata->deform.param_id_in_scene].i != lfodata->deform.val.i);
+
+        if (changed)
+        {
+            bdwLight->dirty = true;
+            bdwBG->dirty = true;
+        }
+    }
+    void drawWaveform(NVGcontext *vg)
+    {
+        auto lfodata = module->lfostorageDisplay;
+        auto storage = module->storage.get();
+
+        tp[lfodata->delay.param_id_in_scene].i = lfodata->delay.val.i;
+        tp[lfodata->attack.param_id_in_scene].i = lfodata->attack.val.i;
+        tp[lfodata->hold.param_id_in_scene].i = lfodata->hold.val.i;
+        tp[lfodata->decay.param_id_in_scene].i = lfodata->decay.val.i;
+        tp[lfodata->sustain.param_id_in_scene].i = lfodata->sustain.val.i;
+        tp[lfodata->release.param_id_in_scene].i = lfodata->release.val.i;
+
+        tp[lfodata->magnitude.param_id_in_scene].i = lfodata->magnitude.val.i;
+        tp[lfodata->rate.param_id_in_scene].i = lfodata->rate.val.i;
+        tp[lfodata->shape.param_id_in_scene].i = lfodata->shape.val.i;
+        tp[lfodata->start_phase.param_id_in_scene].i = lfodata->start_phase.val.i;
+        tp[lfodata->deform.param_id_in_scene].i = lfodata->deform.val.i;
+        tp[lfodata->trigmode.param_id_in_scene].i = lm_keytrigger;
+
+        float susTime = 0.5;
+        bool msegRelease = false;
+        float msegReleaseAt = 0;
+        float lfoEnvelopeDAHDTime =
+            pow(2.0f, lfodata->delay.val.f) + pow(2.0f, lfodata->attack.val.f) +
+            pow(2.0f, lfodata->hold.val.f) + pow(2.0f, lfodata->decay.val.f);
+
+        float totalEnvTime =
+            lfoEnvelopeDAHDTime + std::min(pow(2.0f, lfodata->release.val.f), 4.f) +
+            0.5; // susTime; this is now 0.5 to keep the envelope fixed in gate mode
+
+        float rateInHz = pow(2.0, (double)lfodata->rate.val.f);
+        if (lfodata->rate.temposync)
+            rateInHz *= storage->temposyncratio;
+
+        /*
+         * What we want is no more than 50 wavelengths. So
+         *
+         * totalEnvTime * rateInHz < 50
+         *
+         * totalEnvTime < 50 / rateInHz
+         *
+         * so
+         */
+        totalEnvTime = std::min(totalEnvTime, 50.f / rateInHz);
+
+        std::unique_ptr<LFOModulationSource> tlfo = std::make_unique<LFOModulationSource>();
+        tlfo->assign(storage, lfodata, tp, nullptr, module->surge_ss.get(), module->surge_ms.get(),
+                     module->surge_fs.get(), true);
+        tlfo->attack();
+
+        bool hasFullWave = false, waveIsAmpWave = false;
+
+        bool drawEnvelope = !lfodata->delay.deactivated;
+
+        int minSamples = (1 << 0) * (int)(box.size.x);
+        int totalSamples =
+            std::max((int)minSamples, (int)(totalEnvTime * storage->samplerate / BLOCK_SIZE));
+        float drawnTime = totalSamples * storage->samplerate_inv * BLOCK_SIZE;
+
+        // OK so let's assume we want about 1000 pixels worth tops in
+        int averagingWindow = (int)(totalSamples / 1000.0) + 1;
+
+        float valScale = 100.0;
+        int susCountdown = -1;
+
+        float priorval = 0.f, priorwval = 0.f;
+
+        std::string invalidMessage;
+
+        std::vector<std::pair<float, float>> pathV, eupathV, edpathV;
+
+        for (int i = 0; i < totalSamples; i += averagingWindow)
+        {
+            float val = 0;
+            float wval = 0;
+            float eval = 0;
+            float minval = 1000000, minwval = 1000000;
+            float maxval = -1000000, maxwval = -1000000;
+            float firstval;
+            float lastval;
+
+            for (int s = 0; s < averagingWindow; s++)
+            {
+                tlfo->process_block();
+
+                if (susCountdown < 0 && tlfo->env_state == lfoeg_stuck)
+                {
+                    susCountdown = susTime * storage->samplerate / BLOCK_SIZE;
+                }
+                else if (susCountdown == 0 && tlfo->env_state == lfoeg_stuck)
+                {
+                    tlfo->release();
+                }
+                else if (susCountdown > 0)
+                {
+                    susCountdown--;
+                }
+
+                val += tlfo->get_output(0);
+
+                if (s == 0)
+                {
+                    firstval = tlfo->get_output(0);
+                }
+
+                if (s == averagingWindow - 1)
+                {
+                    lastval = tlfo->get_output(0);
+                }
+
+                minval = std::min(tlfo->get_output(0), minval);
+                maxval = std::max(tlfo->get_output(0), maxval);
+                eval += tlfo->env_val * lfodata->magnitude.get_extended(lfodata->magnitude.val.f);
+            }
+
+            val = val / averagingWindow;
+            wval = wval / averagingWindow;
+            eval = eval / averagingWindow;
+            val = ((-val + 1.0f) * 0.5 * 0.8 + 0.1) * valScale;
+            wval = ((-wval + 1.0f) * 0.5 * 0.8 + 0.1) * valScale;
+            minwval = ((-minwval + 1.0f) * 0.5 * 0.8 + 0.1) * valScale;
+            maxwval = ((-maxwval + 1.0f) * 0.5 * 0.8 + 0.1) * valScale;
+
+            float euval = ((-eval + 1.0f) * 0.5 * 0.8 + 0.1) * valScale;
+            float edval = ((eval + 1.0f) * 0.5 * 0.8 + 0.1) * valScale;
+            float xc = valScale * i / totalSamples;
+
+            if (i == 0)
+            {
+                pathV.emplace_back(xc, val);
+                eupathV.emplace_back(xc, euval);
+                edpathV.emplace_back(xc, edval);
+
+                priorval = val;
+                priorwval = wval;
+            }
+            else
+            {
+                minval = ((-minval + 1.0f) * 0.5 * 0.8 + 0.1) * valScale;
+                maxval = ((-maxval + 1.0f) * 0.5 * 0.8 + 0.1) * valScale;
+                // Windows is sensitive to out-of-order line draws in a way which causes spikes.
+                // Make sure we draw one closest to prior first. See #1438
+                float firstval = minval;
+                float secondval = maxval;
+
+                if (priorval - minval < maxval - priorval)
+                {
+                    firstval = maxval;
+                    secondval = minval;
+                }
+
+                pathV.emplace_back(xc - 0.1 * valScale / totalSamples, firstval);
+                pathV.emplace_back(xc + 0.1 * valScale / totalSamples, secondval);
+
+                priorval = val;
+                eupathV.emplace_back(xc, euval);
+                edpathV.emplace_back(xc, edval);
+            }
+        }
+
+        tlfo->completedModulation();
+        {
+            bool first{true};
+            nvgBeginPath(vg);
+            auto sx = box.size.x / valScale, sy = box.size.y / valScale;
+            for (const auto &[x, y] : eupathV)
+            {
+                if (first)
+                {
+                    nvgMoveTo(vg, sx * x, sy * y);
+                }
+                else
+                {
+                    nvgLineTo(vg, sx * x, sy * y);
+                }
+                first = false;
+            }
+            nvgStrokeColor(vg, style()->getColor(style::XTStyle::PLOT_CURVE));
+            nvgStrokeWidth(vg, 0.5);
+            nvgStroke(vg);
+        }
+        {
+            bool first{true};
+            nvgBeginPath(vg);
+            auto sx = box.size.x / valScale, sy = box.size.y / valScale;
+            for (const auto &[x, y] : edpathV)
+            {
+                if (first)
+                {
+                    nvgMoveTo(vg, sx * x, sy * y);
+                }
+                else
+                {
+                    nvgLineTo(vg, sx * x, sy * y);
+                }
+                first = false;
+            }
+            nvgStrokeColor(vg, style()->getColor(style::XTStyle::PLOT_CURVE));
+            nvgStrokeWidth(vg, 0.5);
+            nvgStroke(vg);
+        }
+        {
+            bool first{true};
+            nvgBeginPath(vg);
+            auto sx = box.size.x / valScale, sy = box.size.y / valScale;
+            for (const auto &[x, y] : pathV)
+            {
+                if (first)
+                {
+                    nvgMoveTo(vg, sx * x, sy * y);
+                }
+                else
+                {
+                    nvgLineTo(vg, sx * x, sy * y);
+                }
+                first = false;
+            }
+            nvgStrokeColor(vg, style()->getColor(style::XTStyle::PLOT_CURVE));
+            nvgStrokeWidth(vg, 1.25);
+            nvgStroke(vg);
+        }
+
+#if 0
+        if (tFullWave)
+        {
+            tFullWave->completedModulation();
+            delete tFullWave;
+        }
+
+        auto at = juce::AffineTransform()
+                      .scale(waveform_display.getWidth() / valScale,
+                             waveform_display.getHeight() / valScale)
+                      .translated(waveform_display.getTopLeft());
+
+        for (int i = -1; i < 2; ++i)
+        {
+            float off = i * 0.4 + 0.5; // so that/s 0.1, 0.5, 0.9
+            float x0 = 0, y0 = valScale * off, x1 = valScale, y1 = valScale * off;
+
+            at.transformPoint(x0, y0);
+            at.transformPoint(x1, y1);
+
+            if (i == 0)
+            {
+                g.setColour(skin->getColor(Colors::LFO::Waveform::Center));
+            }
+            else
+            {
+                g.setColour(skin->getColor(Colors::LFO::Waveform::Bounds));
+            }
+
+            g.drawLine(x0, y0, x1, y1, 1);
+        }
+
+        auto pointColor = skin->getColor(Colors::LFO::Waveform::Dots);
+        int nxd = 61, nyd = 9;
+
+        for (int xd = 0; xd < nxd; xd++)
+        {
+            float normx = 1.f * xd / (nxd - 1) * 0.99;
+            for (int yd = 1; yd < nyd - 1; yd++)
+            {
+                if (yd == (nyd - 1) / 2)
+                    continue;
+
+                float normy = 1.f * yd / (nyd - 1);
+                float dotPointX = normx * valScale, dotPointY = (0.8 * normy + 0.1) * valScale;
+                at.transformPoint(dotPointX, dotPointY);
+                float esize = 1.f;
+                float xoff = (xd == 0 ? esize : 0);
+                g.setColour(pointColor);
+                g.fillEllipse(dotPointX - esize, dotPointY - esize, esize, esize);
+            }
+        }
+
+        if (drawEnvelope)
+        {
+            g.setColour(skin->getColor(Colors::LFO::Waveform::Envelope));
+            g.strokePath(eupath, juce::PathStrokeType(1.f), at);
+
+            if (!isUnipolar())
+            {
+                g.strokePath(edpath, juce::PathStrokeType(1.f), at);
+            }
+        }
+
+        if (drawBeats)
+        {
+            // calculate beat grid related data
+            auto bpm = storage->temposyncratio * 120;
+
+            auto denFactor = 4.0 / tsDen;
+            auto beatsPerSecond = bpm / 60.0;
+            auto secondsPerBeat = 1 / beatsPerSecond;
+            auto deltaBeat = secondsPerBeat / drawnTime * valScale * denFactor;
+
+            int nBeats = std::max(ceil(drawnTime * beatsPerSecond / denFactor), 1.0);
+
+            auto measureLen = deltaBeat * tsNum;
+            int everyMeasure = 1;
+
+            while (measureLen < 4) // a hand selected parameter by playing around with it, tbh
+            {
+                measureLen *= 2;
+                everyMeasure *= 2;
+            }
+
+            for (auto l = 0; l <= nBeats; ++l)
+            {
+                auto xp = deltaBeat * l;
+
+                if (l % (tsNum * everyMeasure) == 0 || nBeats <= tsDen)
+                {
+                    auto soff = 0.0;
+                    if (l > 10)
+                        soff = 0.0;
+                    float vruleSX = xp, vruleSY = valScale * .15, vruleEX = xp,
+                          vruleEY = valScale * .85;
+                    at.transformPoint(vruleSX, vruleSY);
+                    at.transformPoint(vruleEX, vruleEY);
+                    // major beat divisions on the LFO waveform bg
+                    g.setColour(skin->getColor(Colors::LFO::Waveform::Ruler::ExtendedTicks));
+                    g.drawLine(vruleSX, vruleSY, vruleEX, vruleEY, 1);
+
+                    juce::Point<float> sp(xp, valScale * (.01)), ep(xp, valScale * (.1));
+                    sp = sp.transformedBy(at);
+                    ep = ep.transformedBy(at);
+                    // ticks for major beats
+                    if (l % tsNum == 0)
+                        g.setColour(skin->getColor(Colors::LFO::Waveform::Ruler::Ticks));
+                    else
+                        // small ticks for the ruler for nbeats case
+                        g.setColour(skin->getColor(Colors::LFO::Waveform::Ruler::SmallTicks));
+
+                    g.drawLine(sp.getX(), sp.getY(), ep.getX(), ep.getY(), 1.0);
+
+                    char s[TXT_SIZE];
+                    snprintf(s, TXT_SIZE, "%d", l + 1);
+
+                    juce::Point<int> tp(xp + 1, valScale * 0.0);
+                    tp = tp.transformedBy(at);
+                    g.setColour(skin->getColor(Colors::LFO::Waveform::Ruler::Text));
+                    g.setFont(skin->fontManager->lfoTypeFont);
+                    g.drawText(s, tp.x, tp.y, 20, 10, juce::Justification::bottomLeft);
+                }
+                else if (everyMeasure == 1)
+                {
+                    auto sp = juce::Point<float>(xp, valScale * 0.06).transformedBy(at);
+                    auto ep = juce::Point<float>(xp, valScale * 0.1).transformedBy(at);
+
+                    if (l % tsNum == 0)
+                        g.setColour(juce::Colours::black);
+                    else
+                        // small ticks for the ruler
+                        g.setColour(skin->getColor(Colors::LFO::Waveform::Ruler::SmallTicks));
+                    g.drawLine(sp.x, sp.y, ep.x, ep.y, 0.5);
+                }
+            }
+        }
+
+        if (hasFullWave)
+        {
+            if (waveIsAmpWave)
+            {
+                g.setColour(skin->getColor(Colors::LFO::Waveform::GhostedWave));
+                auto dotted = juce::Path();
+                auto st = juce::PathStrokeType(0.3, juce::PathStrokeType::beveled,
+                                               juce::PathStrokeType::butt);
+                float dashLength[2] = {4.f, 2.f};
+                st.createDashedStroke(dotted, deactPath, dashLength, 2, at);
+                g.strokePath(dotted, st);
+            }
+            else
+            {
+                g.setColour(skin->getColor(Colors::LFO::Waveform::DeactivatedWave));
+                g.strokePath(deactPath,
+                             juce::PathStrokeType(0.5f, juce::PathStrokeType::beveled,
+                                                  juce::PathStrokeType::butt),
+                             at);
+            }
+        }
+
+        g.setColour(skin->getColor(Colors::LFO::Waveform::Wave));
+        g.strokePath(
+            path,
+            juce::PathStrokeType(1.f, juce::PathStrokeType::beveled, juce::PathStrokeType::butt),
+            at);
+
+        // lower ruler calculation
+        // find time delta
+        int maxNumLabels = 5;
+        std::vector<float> timeDeltas = {0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0};
+        auto currDelta = timeDeltas.begin();
+        while (currDelta != timeDeltas.end() && (drawnTime / *currDelta) > maxNumLabels)
+        {
+            currDelta++;
+        }
+        float delta = timeDeltas.back();
+        if (currDelta != timeDeltas.end())
+            delta = *currDelta;
+
+        int nLabels = (int)(drawnTime / delta) + 1;
+        float dx = delta / drawnTime * valScale;
+
+        for (int l = 0; l < nLabels; ++l)
+        {
+            float xp = dx * l;
+            float yp = valScale * 0.9;
+            float typ = yp;
+            auto tp = juce::Point<float>(xp + 0.5, typ + 0.5).transformedBy(at);
+            g.setColour(skin->getColor(Colors::LFO::Waveform::Ruler::Text));
+            g.setFont(skin->fontManager->lfoTypeFont);
+            char txt[TXT_SIZE];
+            float tv = delta * l;
+            if (fabs(roundf(tv) - tv) < 0.05)
+                snprintf(txt, TXT_SIZE, "%d s", (int)round(delta * l));
+            else if (delta < 0.1)
+                snprintf(txt, TXT_SIZE, "%.2f s", delta * l);
+            else
+                snprintf(txt, TXT_SIZE, "%.1f s", delta * l);
+            g.drawText(txt, tp.x, tp.y, 30, 10, juce::Justification::topLeft);
+
+            auto sp = juce::Point<float>(xp, valScale * 0.95).transformedBy(at);
+            auto ep = juce::Point<float>(xp, valScale * 0.9).transformedBy(at);
+
+            g.setColour(skin->getColor(Colors::LFO::Waveform::Ruler::Ticks));
+            g.drawLine(sp.x, sp.y, ep.x, ep.y, 1.0);
+        }
+
+        /*
+         * In 1.8 I think we don't want to show the key release point in the simulated wave
+         * with the MSEG but I wrote it to debug and we may change our mind so keeping this code
+         * here
+         */
+        if (msegRelease && false)
+        {
+#if SHOW_RELEASE_TIMES
+            float xp = msegReleaseAt / drawnTime * valScale;
+            was a vstgui Point sp(xp, valScale * 0.9), ep(xp, valScale * 0.1);
+            tf.transform(sp);
+            tf.transform(ep);
+            dc->setFrameColor(juce::Colours::red);
+            dc->drawLine(sp, ep);
+#endif
+        }
+
+        if (warnForInvalid)
+        {
+            g.setColour(skin->getColor(Colors::LFO::Waveform::Wave));
+            g.setFont(skin->fontManager->getLatoAtSize(14, juce::Font::bold));
+            g.drawText(invalidMessage, waveform_display.withTrimmedBottom(30),
+                       juce::Justification::centred);
+        }
+#endif
+    }
 };
 
 LFOWidget::LFOWidget(LFOWidget::M *module) : XTModuleWidget()
@@ -276,6 +781,15 @@ LFOWidget::LFOWidget(LFOWidget::M *module) : XTModuleWidget()
 
     auto tw = LFOTypeWidget::create(screwWidth, module, M::SHAPE);
     addChild(tw);
+
+    auto ww = new LFOWaveform();
+    ww->module = module;
+    ww->box.pos.x = tw->box.pos.x;
+    ww->box.pos.y = tw->box.pos.y + tw->box.size.y;
+    ww->box.size.x = tw->box.size.x;
+    ww->box.size.y = rack::mm2px(ht) - ww->box.pos.y - rack::mm2px(6);
+    ww->setup();
+    addChild(ww);
 
     {
         int col = 0;

--- a/src/LFO.h
+++ b/src/LFO.h
@@ -95,6 +95,7 @@ struct LFO : modules::XTModule
     std::unique_ptr<MSEGStorage> surge_ms;
     std::unique_ptr<FormulaModulatorStorage> surge_fs;
     LFOStorage *lfostorage;
+    LFOStorage *lfostorageDisplay;
 
     modules::ModulationAssistant<LFO, n_lfo_params, RATE, n_mod_inputs, LFO_MOD_INPUT> modAssist;
 
@@ -107,6 +108,7 @@ struct LFO : modules::XTModule
         setupSurgeCommon(NUM_PARAMS, false);
         for (int i = 0; i < MAX_POLY; ++i)
             surge_lfo[i] = std::make_unique<LFOModulationSource>();
+
         surge_ss = std::make_unique<StepSequencerStorage>();
         surge_ms = std::make_unique<MSEGStorage>();
         surge_fs = std::make_unique<FormulaModulatorStorage>();
@@ -114,6 +116,10 @@ struct LFO : modules::XTModule
         lfostorage = &(storage->getPatch().scene[0].lfo[0]);
         lfostorage->delay.deactivated = false;
         lfostorage->trigmode.val.i = lm_keytrigger;
+
+        lfostorageDisplay = &(storage->getPatch().scene[0].lfo[1]);
+        lfostorageDisplay->delay.deactivated = false;
+        lfostorageDisplay->trigmode.val.i = lm_keytrigger;
 
         paramOffsetByID[RATE] = 0;
         paramOffsetByID[SHAPE] = 1;
@@ -250,6 +256,16 @@ struct LFO : modules::XTModule
                     output0[s][i] = output1[s][i];
 
             float ts[3][16];
+
+            {
+                // Setup the display storage
+                auto *par0 = &(lfostorageDisplay->rate);
+                for (int p = RATE; p < LFO_MOD_PARAM_0; ++p)
+                {
+                    auto *par = &par0[paramOffsetByID[p]];
+                    par->set_value_f01(params[p].getValue());
+                }
+            }
 
             for (int c = 0; c < nChan; ++c)
             {

--- a/src/VCO.h
+++ b/src/VCO.h
@@ -695,27 +695,14 @@ template <int oscType> struct VCO : public modules::XTModule
                 }
 
                 oscstorage->wt.current_id = supposedIdx;
-                strncpy(oscstorage->wavetable_display_name, dname.c_str(), 256);
+                strncpy(oscstorage->wavetable_display_name, dname.c_str(), 255);
 
                 oscstorage_display->wt.current_id = supposedIdx;
-                strncpy(oscstorage_display->wavetable_display_name, dname.c_str(), 256);
+                strncpy(oscstorage_display->wavetable_display_name, dname.c_str(), 255);
 
                 wavetableIndex = supposedIdx;
             }
         }
-
-        auto getBool = [](auto *n, auto *str, auto def) {
-            auto v = json_object_get(n, str);
-            if (!v)
-                return def;
-            return json_boolean_value(v);
-        };
-        auto getInt = [](auto *n, auto *str, auto def) -> int {
-            auto v = json_object_get(n, str);
-            if (!v)
-                return def;
-            return json_integer_value(v);
-        };
     }
 };
 

--- a/src/XTModule.h
+++ b/src/XTModule.h
@@ -101,8 +101,8 @@ struct XTModule : public rack::Module
         if (!showedPathsOnce)
         {
             showedPathsOnce = true;
-            INFO("[SurgeXTRack] storage::dataPath = '%s'", storage->datapath.c_str());
-            INFO("[SurgeXTRack] storage::userDataPath = '%s'", storage->userDataPath.c_str());
+            INFO("[SurgeXTRack] storage::dataPath = '%s'", storage->datapath.u8string().c_str());
+            INFO("[SurgeXTRack] storage::userDataPath = '%s'", storage->userDataPath.u8string().c_str());
         }
 
         onSampleRateChange();

